src/util/mapbox.js

```diff
@@ -133,19 +133,122 @@ function formatUrl(obj: UrlObject): string {
     return `${obj.protocol}://${obj.authority}${obj.path}${params}`;
 }
 
-export class TurnstileEvent {
+class TelemetryEvent {
     eventData: { anonId: ?string, lastSuccess: ?number, accessToken: ?string};
-    queue: Array<number>;
-    pending: boolean
+    queue: Array<any>;
     pendingRequest: ?Cancelable;
 
     constructor() {
         this.eventData = { anonId: null, lastSuccess: null, accessToken: config.ACCESS_TOKEN};
         this.queue = [];
-        this.pending = false;
         this.pendingRequest = null;
     }
 
+    fetchEventData() {
+        const isLocalStorageAvailable = storageAvailable('localStorage');
+        const storageKey = `${telemEventKey}:${config.ACCESS_TOKEN || ''}`;
+
+        if (isLocalStorageAvailable) {
+            //Retrieve cached data
+            try {
+                const data = window.localStorage.getItem(storageKey);
+                if (data) {
+                    this.eventData = JSON.parse(data);
+                }
+            } catch (e) {
+                warnOnce('Unable to read from LocalStorage');
+            }
+        }
+    }
+
+    saveEventData() {
+        const isLocalStorageAvailable = storageAvailable('localStorage');
+        const storageKey = `${telemEventKey}:${config.ACCESS_TOKEN || ''}`;
+
+        if (isLocalStorageAvailable) {
+            try {
+                window.localStorage.setItem(storageKey, JSON.stringify(this.eventData));
+            } catch (e) {
+                warnOnce('Unable to write to LocalStorage');
+            }
+        }
+
+    }
+
+    processRequests() {}
+
+    queueRequest(date: number | {id: number, timestamp: number}) {
+        this.queue.push(date);
+        this.processRequests();
+    }
+}
+
+export class MapLoadEvent extends TelemetryEvent {
+    eventData: { anonId: ?string, lastSuccess: ?number, accessToken: ?string};
+    queue: Array<{ id: number, timestamp: number}>;
+    pendingRequest: ?Cancelable;
+    +success: {[number]: boolean};
+
+    constructor() {
+        super();
+        this.success = {};
+    }
+
+    postMapLoadEvent(tileUrls: Array<string>, mapId: number) {
+        //Enabled only when Mapbox Access Token is set and a source uses
+        // mapbox tiles.
+        if (config.ACCESS_TOKEN &&
+            Array.isArray(tileUrls) &&
+            tileUrls.some(url => isMapboxHTTPURL(url))) {
+            this.queueRequest({id: mapId, timestamp: Date.now()});
+        }
+    }
+
+    processRequests() {
+        if (this.pendingRequest || this.queue.length === 0) return;
+        const {id, timestamp} = this.queue.shift();
+
+        // Only one load event should fire per map
+        if (id && this.success[id]) return;
+
+        if (!this.eventData.anonId) {
+            this.fetchEventData();
+        }
+
+        if (!validateUuid(this.eventData.anonId)) {
+            this.eventData.anonId = uuid();
+        }
+
+        const eventsUrlObject: UrlObject = parseUrl(config.EVENTS_URL);
```

Comment:
Code to send the request is common for all telemetry events, maybe it could be pulled into `TelemetryEvent` with something like:
```
sendTelemetryEvent(eventId, timestamp, { ..other event properties }, callback).
```
It would also ensure that all events correctly save event data and process the queue